packages/babel-plugin-proposal-class-properties/src/index.js

```diff
@@ -98,9 +112,9 @@ export default function(api, options) {
 
         let ref;
 
-        if (path.isClassExpression() || !path.node.id) {
+        if (isClassExpression || !path.node.id) {
           nameFunction(path);
-          ref = path.scope.generateUidIdentifier("class");
+          ref = path.node.id || path.scope.generateUidIdentifier("class");
```

Comment:
You should always generate a new identifier, otherwise it won't work in cases like https://babeljs.io/repl/build/6407/#?babili=false&browsers=&build=&builtIns=false&code_lz=DYUwLgBAghC8ECYDcAoFAzAdgCgMbAEMBnI6CAbxQgiLALAEtcIAHAJwHsW5FUBfAJRIgA&debug=false&circleciRepo=&evaluate=false&lineWrap=false&presets=stage-0&prettier=false&targets=&version=7.0.0-beta.36%2Bpr.6963 (`A` gets declared twice).

You can do something like
```js
ref = path.scope.generateUidIdentifier(path.node.id ? path.node.id.name : "class");
```

Reply:
Good catch. Need to dig into this after work, as I have a feeling that it will break inlining `ClassExpression`s into a class IIFE [here](https://github.com/babel/babel/blob/32f20f85815dce5e90dae5db06d1ae2716bd0f4b/packages/babel-plugin-transform-classes/src/index.js#L67), so I will have to check this and probably fix.

Reply:
Made a quick test and as suspected - it breaks some cases. Maybe you would have a possible solution/approach in mind to fix this.

The problem is that this:
```js
() => class Bar {
  method(){}
  static fn = () => 'static';
};
```
gets transformed into with class-properties transform
```js
() => {
  var _Bar;
  return _Bar =
  /*#__PURE__*/
  (() => {
    let _Bar = class Bar {
      method() {}
    };
    _Bar.fn = () => 'static';
    return _Bar;
  })();
}
```
which gets later transformed with `transform-classes` into:
```js
() => {
  var _Bar;
  return _Bar =
  /*#__PURE__*/
  (() => {
    function Bar() {}
    var _proto = Bar.prototype;
    _proto.method = function method() {};
    _Bar.fn = () => 'static';
    return _Bar;
  })();
};
```

The problem lies in the fact that the mentioned [inlining](https://github.com/babel/babel/blob/32f20f85815dce5e90dae5db06d1ae2716bd0f4b/packages/babel-plugin-transform-classes/src/index.js#L67) doesn't know that the local class ID is other than that class ID itself.

What comes to my mind is to:
- retrieve this ID - we know for sure in this place that the `parentStatement` (the one that is going to be replaced) is a `VariableDeclaration` containing the correct ID
- rename references to the old ID (constructor's ID) with the retrieved ID (Im not sure how to do this right now, as I'd like to update identifiers in array of nodes and not a path)
- actually do not remove `parentStatement`, but just replace it's `init` with a constructor (from the nodes that we want to insert here), to keep a local ID and a class proper name

Reply:
Probably the right fix is avoiding adding `var [name]` for class expressions.

e.g.
https://babeljs.io/repl/build/6407/#?babili=false&browsers=&build=&builtIns=false&code_lz=BQYwNghgzlAECCsDeAoWsoBcKYJYlgDMB7Y2AXlgCYBuFAXwEog&debug=false&circleciRepo=&evaluate=false&lineWrap=false&presets=stage-0&prettier=false&targets=&version=7.0.0-beta.36%2Bpr.6963

Input:
```js
(class A {
  static foo = 2;
})
```

Expected output
```js
/*#__PURE__*/
(() => {
  let A = class A {};
  Object.defineProperty(A, "foo", {
    configurable: true,
    enumerable: true,
    writable: true,
    value: 2
  });
  return A;
})();

// Which then becomes

/*#__PURE__*/
(function () {
  var A = function A() {
    _classCallCheck(this, A);
  };

  Object.defineProperty(A, "foo", {
    configurable: true,
    enumerable: true,
    writable: true,
    value: 2
  });
  return A;
})();
```

Actual output
```js
var A;

A =
/*#__PURE__*/
(() => {
  let A = class A {};
  Object.defineProperty(A, "foo", {
    configurable: true,
    enumerable: true,
    writable: true,
    value: 2
  });
  return A;
})();

// Which then becomes

var A;

A =
/*#__PURE__*/
function () {
  var A = function A() {
    _classCallCheck(this, A);
  };

  Object.defineProperty(A, "foo", {
    configurable: true,
    enumerable: true,
    writable: true,
    value: 2
  });
  return A;
}();
```